### ISSUE TITLE
2d detection boxes randomly dropping out, varies image to image, within dataset (border collision?)

### ISSUE THREAD

**Issue by lessw2020** [Unity-Technologies/com.unity.perception#282]
due to some urgent changes, I've had to reconfigure our 2d bounding box layout.  
Somehow that has triggered a problem whereby depending on the image, some of the expected bounding boxes will randomly not be captured.

The key change seems to be is that we now have a lot of boxes with borders that overlap/crowd each other.  I also have turned off the object itself from making a label and it's now 100% invisible quads for all boxes.  
I've carefully adjusted z order, and tried to tighten the borders as much as possible to avoid overlap, and also upgraded to the latest perception version (mar 24).  

Here's some examples of the problem - there should be 6 labelled boxes, 1 of each type, per image like this;
![allwindows](https://user-images.githubusercontent.com/46302957/113228325-76e63b80-9249-11eb-9dc2-0efba1a996bb.JPG)

But as you move through the dataset, various boxes randomly disappear, for no obvious reason - here the small read window is gone even though it is highest on the z stack meaning it shouldn't have any conflict:
![read_window_gone](https://user-images.githubusercontent.com/46302957/113228356-8796b180-9249-11eb-94e2-7e6971c6f98a.JPG)

and then here the bottom window is now gone:
![bottom_gone](https://user-images.githubusercontent.com/46302957/113228377-9715fa80-9249-11eb-875b-eae81afc8fb3.JPG)

Any insight would be appreciated.

Note that if I remove any boxes that could touch, then all is well:
![noborder_noproblem](https://user-images.githubusercontent.com/46302957/113228586-0b509e00-924a-11eb-8256-d097280015f8.JPG)

 I did review the captures.json as well to make sure this wasn't just a bb display issue...the display matches that the boxes were omitted in the capture.

**Comment by lessw2020** [comment]
As a workaround, I'm just going to over-generate and write a script to inspect each of the images and pull only those that have the proper amount of labels so we can still make a solid training set.  Labelling seems to be correct about 45% of the time overall, and 85% if I don't use the small read window.

**Comment by mkamalza** [comment]
Hi Less,

I tried to replicate this issue with a setup similar to yours. Unfortunately, I haven't been able to. I suspect the edges of the example objects I set up do not align the same way yours do. I ca suggest one debugging path that could help further investigate this.

Since you are comfortable with C# code, I suggest cloning the master branch of the Perception repo and then adding the package to your project from the cloned repo instead of the Github URL.

Once this is set up, add these lines to the `OnRenderedObjectInfosCalculated` method of `BoundingBoxLabeler.cs`. On line 147, add:
```C#
if (renderedObjectInfos.Length != 6)
                Debug.Break();
```

`rendererdObjectsInfos` contains information on all the bounding boxes that were detected in the image, so if there are fewer than 6, you know something is wrong. `Debug.Break()` causes the editor to pause on this frame, so that you can investigate the object in the Scene view and see if you can find any visible issues with the z order or the forward direction of the invisible surfaces. Note that if you are using transparent quads and the quads are not facing the camera (perhaps they get rotated somehow), they will be invisible to it.

**Comment by lessw2020** [comment]
Hi @mkamalza - thanks for the info and help!
I will get this setup and debug in.
Since we are under pressure time wise, for now I did just make a filter tool to screen out and remove images that don't have all the requisite labels so it's not blocking us per se.
Debugging in will be much faster than my trial and error yesterday so appreciate the help and will post what I find.
Thanks!
![image_percentage](https://user-images.githubusercontent.com/46302957/113364843-69908600-9309-11eb-9b5f-57dd0c9d2b72.JPG)

**Comment by mkamalza** [comment]
Glad you were able to bypass the issue for now @lessw2020.
Whenever you find some time to debug the issue, we would be curious to hear about your findings in order to help us track down the underlying issue. Thanks!

**Comment by lessw2020** [comment]
Hi @mkamalza  - 
I was able to setup and debug in today.
I could not find anything expressly off, but I did find that simply deleting and remaking the problem quad appears to have fixed it.
After that no matter how I sized or height stacked it I did not see any dropout.  (caveat is I'll need to make a 3K+ size dataset to fully verify but it's is at least much improved and likely fixed).

Note that it was helpful to be able to debug directly as I added some debug log items that help me to monitor right in the console during generation which is helpful.
I'll make a much larger dataset shortly to confirm and post results, but so far looking much, much better.

**Comment by lessw2020** [comment]
I did a 500+ dataset as a faster check, 100% had correct labels at least per label count.
Unfortunately there was no clear isolation of the core problem, but the fix was to delete the quads that were having issues, and remake them from scratch, as after that no more issue.

**Comment by mkamalza** [comment]
Thank you for the update. I'm glad the issue was fixed for now. Do let us know if it pops up again please, we would very much appreciate it!
